# PCA

`VCF=<dir>/holi11.SNP.maxmiss80`

The VCF contains all individuals from across the 11 populations. It's already been filtered by a max missing of 20% (i.e. 0.8 in vcftools)

Sanity check:

1. Check for individual level missingness in the VCF:

vcftools --vcf ${VCF}.vcf --missing-indv --out ${VCF}.missingness

2. Check for coverage:

vcftools --vcf ${VCF}.vcf --depth --out ${VCF}.depth.individuals

3. Check for depth across sites:

vcftools --vcf ${VCF}.vcf --site-mean-depth --out ${VCF}.depth.sites

#### plot all of the above and choose thresholds

4. Filter for depth and remove individuals with high missing data:

```
vcftools --vcf ${VCF}.vcf --min-alleles 2 --max-alleles 2 \
--minGQ 30 --minDP 3 --max-meanDP 80 --max-missing 0.8 \
--min-mac 3 \
--remove $popmap \
--recode --recode-INFO-all --out ${VCF}.filtered
```

#### minDP and max-mean-DP is based on the plots generated by exploring missingness, individual depth, and site-level missingness
#### $popmap is a list of individuals with high missing data / low coverage. Exclude these from the PCA
--------------------------------------------------
## Linkage pruning

#### Make a vcf with IDs to use in plink
`bcftools annotate --set-id +'%CHROM\_%POS' ${VCF}.filtered > ${VCF}_IDs.vcf`

#### Prune SNPs for LD
`plink --vcf ${VCF}_IDs.vcf --out ${VCF}_out_pruned --indep-pairwise 50 5 0.2 --allow-extra-chr`

#### Make Pruned Data
`plink --vcf ${VCF}_IDs.vcf --extract ${VCF}_out_pruned.prune.in --make-bed --out ${VCF}_out_NoLD --allow-extra-chr`

#### Run PCA
`plink --bfile ${VCF}_out_NoLD --pca --out ${VCF}_out_NoLD_PCA --allow-extra-chr`

